stl/inc/algorithm

```diff
@@ -961,6 +941,321 @@ namespace ranges {
+
+                for (;;) { // trim matching suffixes
+                    --_Final1;
+                    --_Final2;
+                    if (!_STD invoke(_Pred, _STD invoke(_Proj1, *_Final1), _STD invoke(_Proj2, *_Final2))) { // mismatch
+                        ++_Final1;
+                        ++_Final2;
+                        break;
+                    }
+
+                    if (--_Count == 1) {
+                        return false; // initial elements still do not match
+                    }
+                }
+                // If we get here, _Count > 1, initial elements do not match, and final elements do not match.
+
+                return _Match_counts(_STD move(_First1), _STD move(_Final1), _STD move(_First2), _STD move(_Final2),
+                    _Pred, _Proj1, _Proj2);
+            } else {
+                return _Match_counts(_STD move(_First1), _STD move(_Last1), _STD move(_First2), _STD move(_Last2),
+                    _Pred, _Proj1, _Proj2);
+            }
+        }
+
+        template <class _It1, class _Se1, class _It2, class _Se2, class _Pr, class _Pj1, class _Pj2>
+        _NODISCARD static constexpr bool _Is_permutation_unsized(
+            _It1 _First1, _Se1 _Last1, _It2 _First2, _Se2 _Last2, _Pr _Pred, _Pj1 _Proj1, _Pj2 _Proj2) {
+            for (;; ++_First1, (void) ++_First2) { // trim matching prefixes
+                if (_First1 == _Last1) { // first range is a prefix of second
+                    return _First2 == _Last2;
+                } else if (_First2 == _Last2) { // second range is a proper prefix of first
+                    return false;
+                }
+
+                if (!_STD invoke(_Pred, _STD invoke(_Proj1, *_First1), _STD invoke(_Proj2, *_First2))) { // mismatch
+                    break;
+                }
+            }
+            // If we get here, initial elements do not match.
+
+            // determine final iterator values and validate lengths
+            auto _Final1 = _First1;
+            auto _Final2 = _First2;
+            for (;;) {
+                ++_Final1;
+                ++_Final2;
+                if (_Final1 == _Last1) {
+                    if (_Final2 == _Last2) {
+                        break; // equal lengths
+                    }
+
+                    return false; // different lengths; not a permutation
+                } else if (_Final2 == _Last2) {
+                    return false; // ditto different lengths
+                }
+            }
+            // If we get here, initial elements do not match and ranges have equal lengths.
+
+            if constexpr (bidirectional_iterator<_It1> && bidirectional_iterator<_It2>) {
+                for (;;) { // trim matching suffixes
+                    if (--_Final1 == _First1) {
+                        return false; // initial elements still do not match
+                    }
+                    --_Final2; // since ranges have equal lengths, _Final2 cannot equal _First2

```

Comment:
Should there be a blank line following this `if` statement?  I know we separate non-chained control flow with blank lines, but seems like there is precedent elsewhere to have a blank line separating an if statement from following statements.  
(Yes, I do realize that blank lines seem to be almost all I care about in this review 😆)

Reply:
If anything, there should be a blank line before the following `if` statement; I'd like to keep these decrements close together to emphasize that the iterators move in lock-step.

Reply:
How about:
```suggestion
                    }
                    --_Final2; // since ranges have equal lengths, _Final2 cannot equal _First2

```